### PR TITLE
MAGN-8159 Library TreeView

### PR DESCRIPTION
**Pull request by ramramps**
### Purpose

This PR is to convert the List view to Tree view (for Library Browse). The Prototype was created for GunSlinger, and the design is approved by UX.  

Here is the new design : 

capture1

1. The header items (Create, Action, Query) were moved vertically, aligned with List items.
2. If the List has only one item, the header items will have only one character displayed (Look at A in the picture above)

Here is the design for a package (Multiple levels)

capture2

This PR has the below tasks
http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8159
 
### Declarations

Check these if you believe they are true

- [X] The code base is in a better state after this PR
- [X] The level of testing this PR includes is appropriate

### Reviewers

@pboyer

### CODE REVIEW

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 1]
```diff
@@ -2013,7 +2020,8 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
             var incomingString = value as string;
 
             if (string.IsNullOrEmpty(incomingString))
-                throw new ArgumentException("value string should not be empty.");
+                //throw new ArgumentException("value string should not be empty.");
```
Comment (pboyer) [inline src/DynamoCoreWpf/UI/Converters.cs]: Please use curly brackets as suggested in the style guide.

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 2]
```diff
@@ -2341,4 +2349,292 @@ public object ConvertBack(object value, Type targetType, object parameter, Cultu
...
+                            if (view.Name == "CategoryTreeView")
+                            {
+                                gotParentTree = true;
+                            }
+                        }
+                    }
+                }
+
+                var left = VerLnMargin.Left + 10;
+                var right = (expanderMargin.Right*2) + 5;
```
Comment (pboyer) [inline src/DynamoCoreWpf/UI/Converters.cs]: Please replace all magic numbers with `const` fields - generally there should be no integers in line in your code.

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 3]
```diff
@@ -2341,4 +2349,292 @@ public object ConvertBack(object value, Type targetType, object parameter, Cultu
...
+        {
+            public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
+            {
+                Thickness margin = (Thickness)value;
+                int bottom = int.Parse(parameter.ToString());
+
+                //If the margin is not set
+                if (margin.Right == 0)
+                    return new Thickness(-10, 0, 0, 0);
+
```
Comment (pboyer) [inline src/DynamoCoreWpf/UI/Converters.cs]: Please use curly braces.

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 4]
```diff
@@ -2341,4 +2349,292 @@ public object ConvertBack(object value, Type targetType, object parameter, Cultu
...
+        /// </summary>
+        public class TreeViewMarginCheck : IValueConverter
+        {
+            public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
+            {
+                Thickness margin = (Thickness)value;
+                 
+                //If the margin is not set
+                if (margin.Right == 0)
+                        return false;
```
Comment (pboyer) [inline src/DynamoCoreWpf/UI/Converters.cs]: Curly braces.

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 5]
```diff
@@ -2341,4 +2349,292 @@ public object ConvertBack(object value, Type targetType, object parameter, Cultu
...
+        }
+
+        /// <summary>
+        /// This controls the extra margin that is drawn even if the margin is not set
+        /// </summary>
+        public class TreeViewMarginCheck : IValueConverter
+        {
+            public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
+            {
+                Thickness margin = (Thickness)value;
```
Comment (pboyer) [inline src/DynamoCoreWpf/UI/Converters.cs]: Use `as` and do a null check - it's faster and doesn't throw an exception.

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 6]
```diff
@@ -2341,4 +2349,292 @@ public object ConvertBack(object value, Type targetType, object parameter, Cultu
...
+                // margin should be -25,0,0,0.
+                if (nestedMargin != null && nestedMargin.Left > 5 && nestedMargin.Right > 0)
+                {
+                    var left = nestedMargin.Right - nestedMargin.Left;
+                    if(left < -5)
+                        //-45,0,0,0 is very close to expander. so move the content a bit.
+                        return new Thickness(left + 5 , 0, 0, 0);
+                    else
+                    {
+                        return new Thickness(-25, 0, 0, 0);
```
Comment (pboyer) [inline src/DynamoCoreWpf/UI/Converters.cs]: Please replace all magic numbers with `const` fields.

File: src/DynamoCoreWpf/UI/Converters.cs (hunk 5 above)
Comment (ramramps) [inline src/DynamoCoreWpf/UI/Converters.cs]: But Thickness is a non nullable value type. So, as cannot be used. But I will do a null check.

File: src/DynamoCoreWpf/ViewModels/Search/BrowserItemViewModel.cs [hunk 7]
```diff
@@ -559,14 +575,20 @@ public ClassInformationViewModel ClassDetails
                 if (classDetails == null && IsClassButton)
                 {
                     classDetails = new ClassInformationViewModel();
-                    classDetails.IsRootCategoryDetails = true;
-                    classDetails.PopulateMemberCollections(this);
+                    classDetails.IsRootCategoryDetails = false;
+                    //classDetails.PopulateMemberCollections(this);
+                    //if (listDetails == null && listDetails.Contains(classDetails))
+                    //    listDetails = new List<ClassInformationViewModel>();
+                    //listDetails.Add(classDetails);
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Search/BrowserItemViewModel.cs]: Let's clean up the commented out code.

File: src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs [hunk 8]
```diff
@@ -243,7 +243,8 @@ private void InitializeCore()
             DefineFullCategoryNames(LibraryRootCategories, "");
             InsertClassesIntoTree(LibraryRootCategories);
 
-            ChangeRootCategoryExpandState(BuiltinNodeCategories.GEOMETRY_CATEGORY, true);
+            //TASK : MAGN 8159 - Do not Expand Geometry by Default.
+            //ChangeRootCategoryExpandState(BuiltinNodeCategories.GEOMETRY_CATEGORY, true);
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs]: Wait, is this already closed? Let's just delete the code.

**Pull request merged**